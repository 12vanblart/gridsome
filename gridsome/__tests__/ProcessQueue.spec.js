const path = require('path')
const ProcessQueue = require('../lib/utils/ProcessQueue')

beforeEach(() => (ProcessQueue.uid = 0))

test('generate srcset for image', async () => {
  const filePath = path.resolve(__dirname, 'assets/1000x600.png')
  const config = { assetsDir: 'assets', maxImageWidth: 1000 }
  const queue = new ProcessQueue({ config })

  const result = await queue.add(filePath)

  expect(result.type).toEqual('png')
  expect(result.filePath).toEqual(filePath)
  expect(result.src).toEqual('/assets/static/1000x600-1000.png')
  expect(result.sizes).toEqual('(max-width: 1000px) 100vw, 1000px')
  expect(result.dataUri).toMatchSnapshot()
  expect(result.imageHTML).toMatchSnapshot()
  expect(result.noscriptHTML).toMatchSnapshot()
  expect(result.sets).toHaveLength(2)
  expect(result.srcset).toHaveLength(2)
  expect(result.sets[0].src).toEqual('/assets/static/1000x600-480.png')
  expect(result.sets[0].width).toEqual(480)
  expect(result.sets[0].height).toEqual(288)
  expect(result.sets[0].type).toEqual('png')
  expect(result.sets[1].src).toEqual('/assets/static/1000x600-1000.png')
  expect(result.sets[1].width).toEqual(1000)
  expect(result.sets[1].height).toEqual(600)
  expect(result.sets[1].type).toEqual('png')
  expect(result.srcset[0]).toEqual('/assets/static/1000x600-480.png 480w')
  expect(result.srcset[1]).toEqual('/assets/static/1000x600-1000.png 1000w')
})

test('resize image by width attribute', async () => {
  const filePath = path.resolve(__dirname, 'assets/1000x600.png')
  const config = { assetsDir: 'assets', maxImageWidth: 1000 }
  const queue = new ProcessQueue({ config })

  const result = await queue.add(filePath, { width: 300 })

  expect(result.src).toEqual('/assets/static/1000x600-300.png')
  expect(result.sizes).toEqual('(max-width: 300px) 100vw, 300px')
  expect(result.dataUri).toMatchSnapshot()
  expect(result.sets).toHaveLength(1)
  expect(result.srcset).toHaveLength(1)
  expect(result.sets[0].src).toEqual('/assets/static/1000x600-300.png')
  expect(result.sets[0].width).toEqual(300)
  expect(result.sets[0].height).toEqual(180)
  expect(result.srcset[0]).toEqual('/assets/static/1000x600-300.png 300w')
})

test('disable blur filter', async () => {
  const filePath = path.resolve(__dirname, 'assets/1000x600.png')
  const config = { assetsDir: 'assets', maxImageWidth: 1000 }
  const queue = new ProcessQueue({ config })

  const result = await queue.add(filePath, { blur: '0' })

  expect(result.dataUri).toMatchSnapshot()
})

test('respect config.maxImageWidth', async () => {
  const filePath = path.resolve(__dirname, 'assets/1000x600.png')
  const config = { assetsDir: 'assets', maxImageWidth: 600 }
  const queue = new ProcessQueue({ config })

  const result = await queue.add(filePath)

  expect(result.src).toEqual('/assets/static/1000x600-600.png')
  expect(result.sets).toHaveLength(2)
  expect(result.srcset).toHaveLength(2)
  expect(result.sets[0].src).toEqual('/assets/static/1000x600-480.png')
  expect(result.sets[1].src).toEqual('/assets/static/1000x600-600.png')
  expect(result.srcset[0]).toEqual('/assets/static/1000x600-480.png 480w')
  expect(result.srcset[1]).toEqual('/assets/static/1000x600-600.png 600w')
})

test('do not resize if image is too small', async () => {
  const filePath = path.resolve(__dirname, 'assets/350x250.png')
  const config = { assetsDir: 'assets', maxImageWidth: 600 }
  const queue = new ProcessQueue({ config })

  const result = await queue.add(filePath, { width: 600 })

  expect(result.src).toEqual('/assets/static/350x250-350.png')
  expect(result.sets).toHaveLength(1)
  expect(result.srcset).toHaveLength(1)
})

test('get url for server in serve mode', async () => {
  const relPath = 'assets/1000x600.png'
  const absPath = path.resolve(__dirname, relPath)
  const config = { assetsDir: 'assets', maxImageWidth: 500 }
  const queue = new ProcessQueue({ config, context: __dirname })
  const mode = process.env.GRIDSOME_MODE

  process.env.GRIDSOME_MODE = 'serve'

  const result = await queue.add(absPath)

  process.env.GRIDSOME_MODE = mode

  expect(result.src).toEqual(`/static/${relPath}?width=500`)
})

test('get queue values', async () => {
  const filePath = path.resolve(__dirname, 'assets/1000x600.png')
  const config = { assetsDir: 'assets', maxImageWidth: 1000 }
  const queue = new ProcessQueue({ config })

  await queue.add(filePath)
  const values = queue.queue

  expect(values).toHaveLength(2)
})

test('disable lazy loading', async () => {
  const filePath = path.resolve(__dirname, 'assets/1000x600.png')
  const config = { assetsDir: 'assets', maxImageWidth: 1000 }
  const queue = new ProcessQueue({ config })

  const result = await queue.add(filePath, { immediate: true })

  expect(result.imageHTML).toMatchSnapshot()
  expect(result.noscriptHTML).toEqual('')
})

test('skip srcset and dataUri', async () => {
  const filePath = path.resolve(__dirname, 'assets/1000x600.png')
  const config = { assetsDir: 'assets', maxImageWidth: 1000 }
  const queue = new ProcessQueue({ config })

  const result = await queue.add(filePath, { srcset: false })

  expect(result.srcset).toBeUndefined()
  expect(result.dataUri).toBeUndefined()
  expect(result.sizes).toBeUndefined()
  expect(result.cacheKey).toBeUndefined()
  expect(result.imageHTML).toMatchSnapshot()
})

test('skip missing files', async () => {
  const filePath = path.resolve(__dirname, 'assets/1000x600-missing.png')
  const config = { assetsDir: 'assets', maxImageWidth: 1000 }
  const queue = new ProcessQueue({ config })

  const result = await queue.add(filePath, { srcset: false })

  expect(result.src).toBeNull()
  expect(result.dataUri).toBeNull()
  expect(result.cacheKey).toBeNull()
  expect(result.sets).toHaveLength(0)
  expect(result.srcset).toHaveLength(0)
  expect(result.sizes.width).toBeNull()
  expect(result.sizes.height).toBeNull()
})
