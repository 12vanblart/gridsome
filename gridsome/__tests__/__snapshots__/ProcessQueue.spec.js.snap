// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`disable blur filter 1`] = `"data:image/svg+xml,%3csvg fill='none' viewBox='0 0 1000 600' xmlns='http://www.w3.org/2000/svg' xmlns:xlink='http://www.w3.org/1999/xlink'%3e%3c/svg%3e"`;

exports[`generate srcset for image 1`] = `"data:image/svg+xml,%3csvg fill='none' viewBox='0 0 1000 600' xmlns='http://www.w3.org/2000/svg' xmlns:xlink='http://www.w3.org/1999/xlink'%3e%3cfilter id='__svg-blur-0'%3e%3cfeGaussianBlur in='SourceGraphic' stdDeviation='40' /%3e%3c/filter%3e%3cimage x='0' y='0' filter='url(%23__svg-blur-0)' width='1000' height='600' xlink:href='data:image/png%3bbase64%2ciVBORw0KGgoAAAANSUhEUgAAAEAAAAAmCAIAAAD7pgrOAAAACXBIWXMAAA7EAAAOxAGVKw4bAAABmElEQVRYw%2b2X3W6CQBBGff/nMF6AIoKCAgVFUJDy90Y9YRNCNe0FTUlJZy7M7PBt/M7MIrJ4n3ksBEAABEAABEAABEAABOBfA5Rl2edFUTxVXkNpXvdOClB2oZI8z1XeJ4/Ho3fW26VCnWVVVSRPe6cG4Iv3%2bz2fYRjquh7HMYmmaUmSBEFAhUQ5w67ymmWZaZqu66Zpul6vPc9DgxI9muFYfh0AN5jA7u12wxA%2bDoeD4zgsbds%2bHo/X65Vl0zRvXWCRHMeWZYHKFgTIELMXpRrIdAB0q65r3/eVD5ZMAx9c2u12p9MJQyChAWm5XF4ul7ZtqXAV9/hmdPAwELZQYTijD9LII8TQccwc%2bPooipwuzuczLkGiQp2uQ7harVgCABgCsMEgQQC22jL6/PzoJsYWneZIbDYbOk2bSe73O54Mw6DCBLgxqKtTTpu32y0YtB8BlxCToJ/6Hhj%2bApZfB576/Hvln36QDVtbdCFPYgEQAAEQAAEQAAEQAAGYAwD/tHln4LUz7YKkX5LwtsQ7xgwmUHyOYYX8A4oOoFMobbRdAAAAAElFTkSuQmCC' /%3e%3c/svg%3e"`;

exports[`resize image by width attribute 1`] = `"data:image/svg+xml,%3csvg fill='none' viewBox='0 0 300 180' xmlns='http://www.w3.org/2000/svg' xmlns:xlink='http://www.w3.org/1999/xlink'%3e%3cfilter id='__svg-blur-0'%3e%3cfeGaussianBlur in='SourceGraphic' stdDeviation='40' /%3e%3c/filter%3e%3cimage x='0' y='0' filter='url(%23__svg-blur-0)' width='300' height='180' xlink:href='data:image/png%3bbase64%2ciVBORw0KGgoAAAANSUhEUgAAAEAAAAAmCAIAAAD7pgrOAAAACXBIWXMAAA7EAAAOxAGVKw4bAAABmElEQVRYw%2b2X3W6CQBBGff/nMF6AIoKCAgVFUJDy90Y9YRNCNe0FTUlJZy7M7PBt/M7MIrJ4n3ksBEAABEAABEAABEAABOBfA5Rl2edFUTxVXkNpXvdOClB2oZI8z1XeJ4/Ho3fW26VCnWVVVSRPe6cG4Iv3%2bz2fYRjquh7HMYmmaUmSBEFAhUQ5w67ymmWZaZqu66Zpul6vPc9DgxI9muFYfh0AN5jA7u12wxA%2bDoeD4zgsbds%2bHo/X65Vl0zRvXWCRHMeWZYHKFgTIELMXpRrIdAB0q65r3/eVD5ZMAx9c2u12p9MJQyChAWm5XF4ul7ZtqXAV9/hmdPAwELZQYTijD9LII8TQccwc%2bPooipwuzuczLkGiQp2uQ7harVgCABgCsMEgQQC22jL6/PzoJsYWneZIbDYbOk2bSe73O54Mw6DCBLgxqKtTTpu32y0YtB8BlxCToJ/6Hhj%2bApZfB576/Hvln36QDVtbdCFPYgEQAAEQAAEQAAEQAAGYAwD/tHln4LUz7YKkX5LwtsQ7xgwmUHyOYYX8A4oOoFMobbRdAAAAAElFTkSuQmCC' /%3e%3c/svg%3e"`;
